### PR TITLE
Fix `rpc::CoreTxIter` logic.

### PR DESCRIPTION
**Pull request by evanlinjin**
### Description

This fixes a bug where `CoreTxIter` attempts to call `listtransactions` immediately after a tx result is filtered (instead of being returned), when in fact, the correct logic will be to pop another tx result.

The new logic also ensures that tx results are returned in chonological order. The test `test_list_transactions` verifies this. We also now ensure that `page_size` is between the range `[0 to 1000]` otherwise an error is returned.

Some needless cloning is removed from `from_config` as well as logging improvements.

### Notes to the reviewers

This is an oversight by me (sorry) for PR #683

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

~* [ ] This pull request breaks the existing API~
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: src/blockchain/rpc.rs [hunk 1]
```diff
@@ -695,81 +701,54 @@ where
...
+                Some(page_size),
+                Some(page_size * page_index),
+                Some(true),
+            )
+        })
+        // take until returned rpc call is empty or until error
+        // TODO: replace with the following when MSRV is 1.57.0:
+        // `.map_while(|res| res.map(|l| if l.is_empty() { None } else { Some(l) }).transpose())`
+        .take_while(|res| {
+            if got_err || res.is_ok() && res.as_ref().unwrap().is_empty() {
```
Comment (afilini) [inline src/blockchain/rpc.rs]: I'm not sure if that works, but maybe you can replace the second condition with something like `matches!(res, Ok(&[]))`.

But if you prefer it this way, that also works for me. Up to you!

**Comment by afilini** [comment]
Can you come up with an example that would break without this PR (i.e. without iterating in chronological order)? You don't have to write it in code right now, just describe it. We can add it later to the blockchain tests so that we can cover more "edge-cases".

**Comment by evanlinjin** [comment]
@afilini the "chronological order" is just a good to have.

The real fix is this:

> There was also a bug where CoreTxIter will attempt to call listtransactions again when a tx result is filtered (the correct logic would be to pop another tx result).

Sorry I should have made it more clear.

It just so happened that the test I wrote for the iterator didn't work initially (and I didn't know why) but then I realized that with different page sizes results were returned in different orders so I decided to change the logic to make things more predictable.

**Comment by evanlinjin** [comment]
@afilini I just updated the PR description and commit message to better describe the intention.

File: src/blockchain/rpc.rs (hunk 1 above)
Comment (evanlinjin) [reply src/blockchain/rpc.rs]: My solution:
```rust
matches!(res, Ok(list) if list.is_empty())
```

**Comment by evanlinjin** [comment]
@afilini Right now the `RpcBlockchain` sync logic assumes that transactions returned by `CoreTxIter` may not in chronological order (this is why we need to calculate `TransactionDetails::received` in a separate loop. In the future we can change this to slightly improve performance (because we are using the new `list_transactions` function).

The chronological nature of transactions returned is more of an "implementation detail" for determining field values of `TransactionDetails` rather than a potential bug. As from my understanding, `Database::iter_txs` does not require transactions to be in chronological order.

**Comment by afilini** [comment]
Ah ok, that's fine.

The other bug was clear to me, instead of an `if let` it should have been a `while let` to iterate until you find a tx that is not filtered. I couldn't figure out why the chronological order was needed, thanks for the explanation!

**Review by afilini — APPROVED** [review]
ACK 74e2c477f124489a2357921ca879cc82e24da5fd
